src/createHistoryMiddleware.js

```diff
@@ -0,0 +1,42 @@
+import ActionTypes from './ActionTypes';
+
+// This is a private action creator not intended to be called by users.
+function locationUpdated(location) {
+  return {
+    type: ActionTypes.LOCATION_UPDATED,
+    payload: location,
+  };
+}
+
+export default function createHistoryMiddleware(protocol) {
+  return function historyMiddleware({ dispatch }) {
+    const dispose = protocol.subscribe((location) => {
```

Comment:
Yeah, there's two ways actions can flow though.

1. The user dispatches a `PUSH` or a `REPLACE` action. This runs through the enhancer middlewares. Then the `historyMiddleware` calls `protocol.transition` and converts this into a `LOCATION_UPDATED` action, which then gets sent down the enhancer/middleware chain.
2. The user hits the back button. The protocol listens for the `popstate` event and calls the subscribed callback. Then the `historyMiddleware` calls `dispatch` (not `next`) on a new `LOCATION_UPDATED` action. This action goes through earlier middlewares, which can then do things like parse the query. It's ignored by `historyMiddleware` the second time around, and then passes through to the store. The same code path is used for `GO` actions as well.